File: app/src/main/java/org/scp/gymlog/room/daos/ExerciseDao.java
package org.scp.gymlog.room.daos;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import org.scp.gymlog.room.entities.ExerciseEntity;

import java.util.List;
import java.util.Optional;

@Dao
public interface ExerciseDao {
    @Query("SELECT * FROM exercise")
    List<ExerciseEntity> getAll();

    @Query("SELECT * FROM exercise WHERE exerciseId = :exerciseId")
    Optional<ExerciseEntity> getById(int exerciseId);

    @Transaction
    @Query("SELECT * FROM exercise")
    List<ExerciseEntity.WithMuscles> getAllWithMuscles();

    @Query("SELECT exerciseId FROM exercise_x_muscle_group WHERE muscleId = :muscleId")
    List<Integer> getExercisesIdByMuscleId(int muscleId);

    @Transaction
    @Query("SELECT * FROM exercise WHERE exerciseId IN (:exerciseIds)")
    List<ExerciseEntity.WithMuscles> getByIds(int... exerciseIds);

    @Insert
    long insert(ExerciseEntity exercise);

    @Insert
    long[] insertAll(ExerciseEntity... exercises);

    @Delete
    int delete(ExerciseEntity... exercises);

    @Update
    void update(ExerciseEntity... exercises);

    @Query("DELETE FROM exercise")
    void clear();
}

File: app/src/main/java/org/scp/gymlog/room/daos/ExerciseDao.kt
package org.scp.gymlog.room.daos

import androidx.room.*
import org.scp.gymlog.room.entities.ExerciseEntity
import org.scp.gymlog.room.entities.ExerciseEntity.WithMusclesAndVariations
import java.util.*

@Dao
interface ExerciseDao {
    @Query("SELECT * FROM exercise")
    fun getAll(): List<ExerciseEntity>

    @Query("SELECT * FROM exercise WHERE exerciseId = :exerciseId")
    fun getById(exerciseId: Int): Optional<ExerciseEntity>

    @Query("SELECT * FROM exercise")
    @Transaction
    fun getAllWithMusclesAndVariations(): List<WithMusclesAndVariations>

    @Query("SELECT exerciseId FROM exercise_x_muscle_group WHERE muscleId = :muscleId")
    fun getExercisesIdByMuscleId(muscleId: Int): List<Int>

    @Insert
    fun insert(exercise: ExerciseEntity): Long

    @Insert
    fun insertAll(vararg exercises: ExerciseEntity): LongArray

    @Delete
    fun delete(vararg exercises: ExerciseEntity): Int

    @Update
    fun update(vararg exercises: ExerciseEntity)

    @Query("DELETE FROM exercise")
    fun clear()
}
